Application/EdFi.Ods.Api/Middleware/EdFiOAuthAuthenticationHandler.cs

```diff
@@ -44,6 +49,29 @@ protected override async Task<AuthenticateResult> HandleAuthenticateAsync()
                     return AuthenticateResult.NoResult();
                 }
 
+                //NOTE: Patch for setting instanceId as IFilterMetadata 'InstanceIdContextFilter' run's too late for Admin and Security database calls.
+                //NOTE: Also, context perhaps has a memory leak as _instanceIdContextProvider.GetInstanceId() returns last api calls instanceId.
+                //This also is the case for schoolYear (i.e _schoolYearContextProvider.GetSchoolYear() returning last api call's schoolYear e.g. 2020, before _schoolYearContextProvider.SetSchoolYear is called).
+                //This is not the same as version 3.4. _schoolYearContextProvider.GetSchoolYear() would return 0 until it was set, rather than returning schoolYear of last api call.
+                /* ------- */
+                if (Request.RouteValues.TryGetValue("instanceIdFromRoute", out object instance))
+                {
+                    // Convert the object value to a string and see if it is empty
+                    var instanceId = instance as string;
+                    if (!string.IsNullOrEmpty(instanceId))
+                    {
+                        const string Pattern = @"^[A-Za-z0-9-]+$";
+                        //check that the character's are allowed
+                        Match match = Regex.Match(instanceId, Pattern);
+                        if (match.Success)
+                        {
+                            // If we're still here, set the context value
+                            _instanceIdContextProvider.SetInstanceId(instanceId);
+                        }
+                    }
+                }
+                /* ------- */
+
```

Comment:
@blmeyers See notes from Mansoor. We need to get InstanceId from the route so we can use the correct Admin/Security database but since InstanceIdContextFilter is called later in the request pipeline, we had to add this code here as well. Hopefully, that is an acceptable solution.